dev/archery/archery/linking.py

```diff
@@ -61,9 +63,84 @@ def list_dependency_names(self):
             names.append(name)
         return names
 
+    def _extract_symbols(self, symbol_info):
+        return [re.search(r'\S+$', line).group() for line in symbol_info if line]
+
+    def _remove_weak_symbols(self, symbol_info):
+        return [line for line in symbol_info if not re.search(r'\s[Ww]\s', line)]
+
+    def _capture_symbols(self, remove_symbol_versions, symbol_info):
+        if remove_symbol_versions:
+            symbol_info = [re.split('@@', line)[0] for line in symbol_info]
+        return symbol_info
+
+    def list_symbols_for_dependency(self, dependency, remove_symbol_versions=False):
+        if dependency == 'linux-vdso.so.1':
+            return []
+        result = _nm.run('-D', dependency, stdout=subprocess.PIPE)
```

Comment:
If you use `nm -P`, you will get the POSIX portability format as described in https://pubs.opengroup.org/onlinepubs/9699919799/utilities/nm.html, which should also be easier to parse (you can just split the output on whitespace).

Reply:
@pitrou I have resolved other comments except this, please check. 
About this one, I agree with you, its way neater, I can see that in my local `nm` version. We have a pretty older version in the CI systems, would it be available even in older versions?

Reply:
Locally testing with `nm -P` , I am getting the following errors
```
nm: /lib/x86_64-linux-gnu/librt.so.1: no symbols
nm: /lib/x86_64-linux-gnu/libdl.so.2: no symbols
nm: /lib/x86_64-linux-gnu/libstdc++.so.6: no symbols
nm: /lib/x86_64-linux-gnu/libm.so.6: no symbols
nm: /lib/x86_64-linux-gnu/libgcc_s.so.1: no symbols
nm: /lib/x86_64-linux-gnu/libc.so.6: no symbols
nm: /lib64/ld-linux-x86-64.so.2: no symbols
```

Reply:
Did you run it on Ubuntu (or Debian)?
Debug information is separated on Ubuntu (or Debian).
(CentOS 7 that is used by java-jars doesn't do it.)

You need to run `nm` against separated debug information file:

```bash
$ sudo apt install -y libc6-dbg
$ LANG=C readelf -n /lib/x86_64-linux-gnu/libc.so.6 | grep "Build ID:"
    Build ID: 3b0430552f7704425a91887df0fcd734bf3c0f6f
$ nm -P /usr/lib/debug/.build-id/3b/0430552f7704425a91887df0fcd734bf3c0f6f.debug
```

Reply:
Aha, I ran it on Ubuntu, yes.

Reply:
So if we use `-P` we have to make sure we run the above and then get symbols right?

Reply:
So probably I may have to debug this on the CI itself and replace `-D` usage with `-P`.

Reply:
You can use `archery docker run -e ARROW_JAVA_BUILD=OFF -e ARROW_JAVA_TEST=OFF java-jni-manylinux-2014`.

Reply:
while building the image, I got the following error

```bash
237.2 error: Failed to download from mirror set
237.2 error: https://github.com/tukaani-project/xz/archive/v5.4.4.tar.gz: curl failed to download with exit code 22
```
https://github.com/tukaani-project/xz/archive/v5.4.4.tar.gz -> Not found

Reply:
There is no difference between `nm` and `nm -P`, at least here on Ubuntu 22.04:
```console
$ LANG=C nm -P /lib/x86_64-linux-gnu/librt.so.1
nm: /lib/x86_64-linux-gnu/librt.so.1: no symbols
$ LANG=C nm /lib/x86_64-linux-gnu/librt.so.1
nm: /lib/x86_64-linux-gnu/librt.so.1: no symbols
```

Reply:
The current code uses nm -D, sorry I am away right now, don't have access to my Ubuntu desktop. I should be able to debug this via archery docker right?

Reply:
Well, you can use `nm -P -D` then.

Reply:
Are you going to try the `nm -P -D` as suggested? or should this be ready for review?

Reply:
No I am going to do what @pitrou said. I am still out on vacation, but I can make the change tomorrow.

Reply:
Changes have been made and checking the CIs, sorry I was away until yesterday couldn't return as expected on Friday.

Reply:
Locally things were fine, but the java-jars crossbow is failing with symbols being detected. Need to check why this is happening?

Reply:
cc @kou @pitrou

Reply:
Could you try `archery docker run ...` again? The xz problem has been fixed.

Reply:
I am trying it... keep you posted.